### PR TITLE
Various portability fixes

### LINKED ISSUE #330: Hadrian assumes path of bash
Trying to build with Hadrian on, for instance, FreeBSD falls over at the first gate with,
```
Linking /usr/home/ben/ghc/hadrian/dist-newstyle/build/hadrian-0.1.0.0/build/hadrian/hadrian ...
/-------------------------------------------------------\
| Run Configure "." (stage = Stage0, package = ghc)     |
|      input: hadrian/cfg/system.config.in (and 2 more) |
|  => output: hadrian/cfg/system.config (and 2 more)    |
\-------------------------------------------------------/
exec: /bin/bash: not found
shakeArgsWith   0.000s    0%                           
Function shake  0.320s   59%  =========================
Database read   0.000s    0%                           
With database   0.000s    0%                           
Running rules   0.220s   40%  =================        
Total           0.542s  100%                           
Error when running Shake build system:
* OracleQ (ConfigKey "host-os")
* hadrian/cfg/system.config
* hadrian/cfg/system.config settings mk/config.h
user error (Development.Shake.cmd, system command failed
Command: /bin/sh -c 'sh configure'
Current directory: /usr/home/ben/ghc
Exit code: 127
Stderr:
exec: /bin/bash: not found
)
```
It's not clear to me where this assumption is coming from.

### PR DESCRIPTION
**Pull request by bgamari**
These are the result of an attempt to build GHC with Hadrian on FreeBSD on amd64. This resolves #330.

### CODE REVIEW

File: src/Oracles/Path.hs [hunk 1]
```diff
@@ -28,7 +28,7 @@ systemBuilderPath builder = case builder of
     Cc  _  Stage0   -> fromKey "system-cc"
     Cc  _  _        -> fromKey "cc"
     -- We can't ask configure for the path to configure!
-    Configure _     -> return "bash configure"
+    Configure _     -> return "sh configure"
```
Comment (snowleopard) [inline src/Oracles/Path.hs]: Should this be `bash ++ " configure"` too?

Comment (bgamari) [inline src/Oracles/Path.hs]: Nope, as I mention in the commit message I don't think there's any reason why we need to use `bash` at all here. `sh` should be perfectly sufficient unless I've missed something.

Comment (snowleopard) [inline src/Oracles/Path.hs]: 👍

File: src/Util.hs [hunk 2]
```diff
@@ -62,10 +62,7 @@ customBuild rs opts target@Target {..} = do
                     top   <- topDirectory
                     cmd [Cwd output] [path] "x" (top -/- input)
 
-            Configure dir -> do
-                -- Inject /bin/bash into `libtool`, instead of /bin/sh
-                let env = AddEnv "CONFIG_SHELL" "/bin/bash"
```
Comment (snowleopard) [inline src/Util.hs]: Removing this breaks the Windows build (the CI failed).

I agree that this bit is mysterious and needs to be better documented and perhaps slightly refactored. Could you inject `bashPath` here and add a TODO for me?

**Comment by bgamari** [comment]
Done.

**Comment by bgamari** [comment]
Uh oh, it looks like this introduces a cycle,
```
/-------------------------------------------------------\
| Run Configure "." (stage = Stage0, package = ghc)     |
|      input: hadrian/cfg/system.config.in (and 2 more) |
|  => output: hadrian/cfg/system.config (and 2 more)    |
\-------------------------------------------------------/
shakeArgsWith   0.000s    0%                           
Function shake  0.270s   64%  =========================
Database read   0.000s    0%                           
With database   0.000s    0%                           
Running rules   0.149s   35%  =============            
Total           0.420s  100%                           
Error when running Shake build system:
* OracleQ (ConfigKey "host-os")
* hadrian/cfg/system.config
* hadrian/cfg/system.config settings mk/config.h
* OracleQ (ConfigKey "host-os")
Build system error - recursion detected:
  Key type:   OracleQ ConfigKey
  Key value:  OracleQ (ConfigKey "host-os")
Rules may not be recursive
```

**Comment by snowleopard** [comment]
@bgamari Yes, it looks like we'll have to hard code `"/bin/bash"` for now. Let's at least make it conditional for running on Windows, i.e. `when (System.os == "mingw32") ...`?

**Comment by bgamari** [comment]
@snowleopard I just dropped the `fixAbsolutePathOnWindows` and it appears to pass on Windows. Any objections to this?

**Comment by snowleopard** [comment]
@bgamari Aha, I see. It's probably fine, so I've merged this. Thank you!

**Pull request merged**